dsgrid/project.py

```diff
@@ -125,8 +164,80 @@ def unload_dataset(self, dataset_id):
             yield dataset
 
     def list_datasets(self):
         return [x.dataset_id for x in self._iter_datasets()]
+
+    @track_timing(timer_stats_collector)
+    def process_query(self, context: QueryContext):
+        self._build_filtered_record_ids_by_dimension_type(context)
+
+        dfs = []
+        for dataset_id in context.model.project.dataset_ids:
+            logger.info("Start processing query for dataset_id=%s", dataset_id)
+            dataset = self.get_dataset(dataset_id)
+            dfs.append(dataset.get_dataframe(context, self._config))
+            logger.info("Finished processing query for dataset_id=%s", dataset_id)
+
+        if not dfs:
+            logger.warning("No data matched %s", context.name)
+            return None
+
+        # All dataset columns need to be in the same order.
+        context.consolidate_dataset_metadata()
+        # TODO: Change this when we support long format.
+        pivoted_columns = sorted(context.get_pivoted_columns())
+        dim_columns = context.get_all_dimension_query_names()
+        time_columns = context.get_dimension_query_names(DimensionType.TIME)
+        dim_columns -= time_columns
+        for col in dim_columns:
+            dimension_type = self._config.get_dimension(col).model.dimension_type
+            if dimension_type == context.get_pivoted_dimension_type():
+                dim_columns.remove(col)
+                break
+        dim_columns = sorted(dim_columns)
+        time_columns = sorted(time_columns)
+        expected_columns = time_columns + pivoted_columns + dim_columns
+        for dataset_id, i in zip(context.model.project.dataset_ids, range(len(dfs))):
+            remaining = sorted(set(dfs[i].columns).difference(expected_columns))
+            final_columns = time_columns + dim_columns + pivoted_columns + remaining
+            missing = context.get_pivoted_columns(dataset_id=dataset_id).difference(dfs[i].columns)
+            for column in missing:
+                dfs[i] = dfs[i].withColumn(column, F.lit(None).cast(DoubleType()))
+            dfs[i] = dfs[i].select(*final_columns)
+
+        main_df = dfs[0]
+        if len(dfs) > 1:
+            for df in dfs[1:]:
+                if df.columns != main_df.columns:
+                    raise Exception(
+                        f"BUG: dataset columns must match. main={main_df.columns} new={df.columns}"
+                    )
+                main_df = main_df.union(df)
+
+        return main_df
+
+    def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+        record_ids = {}
+        base_dim_to_query_name_mapping = self._config.get_base_dimension_to_query_name_mapping()
+
+        for dim_filter in context.model.project.dimension_filters:
+            dim_type = dim_filter.dimension_type
+            query_name = dim_filter.dimension_query_name
+            df = self._config.get_dimension_records(query_name)
+            df = dim_filter.apply_filter(df).select("id")
```

Comment:
I feel like you had a filter model that defaults to id col? If so, should select be based on input for that?